Use the requirement signature in the archetype builder.

huonw opened the pull request:
This allows the archetype builder to avoid traversing all of a protocol, and is even a bit lazier.

huonw committed:
[AST] Store and serialize a protocol's requirement signature.

This is the "canonical" representation of the type-level requirements of
a protocol, and we intend to use it pervasively in the compiler.

huonw committed:
[AST] Use the requirement signature in the ArchetypeBuilder.

The requirement signature is far more compact than looking at all the members.

huonw committed:
[AST] Concretize nested potential archetypes on demand.

When using the requirement signature in the archetype builder, not all
nested types are resolved (i.e. added to their parent PA's NestedTypes
map) when a protocol conformance is added. Equating a PA to a concrete
type would previously rely on the NestedTypes map being complete, to
recursively equate those nested types to the type implied by their
parent being concrete. The lazier resolution now also includes resolving
nested PA's to concrete types, if their parents are concrete.

DougGregor commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2657,6 +2657,8 @@ void Serializer::writeDecl(const Decl *D) {
 
     writeGenericParams(proto->getGenericParams());
+    writeGenericRequirements(
+        proto->getRequirementSignature()->getRequirements(), DeclTypeAbbrCodes);
```
Please bump the module format version.

huonw replied:
This is `VERSION_MINOR` in ModuleFormat.h?

DougGregor replied:
Yes

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4372,6 +4372,8 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
 
     TC.checkDeclAttributes(PD);
 
+    PD->computeRequirementSignature();
```
Hmm... should this go into `TypeChecker::validateDecl()`? It seems like we'd want the requirement signature most of the time that we look at the `ProtocolDecl`.

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1013,52 +1076,66 @@ bool ArchetypeBuilder::addConformanceRequirement(PotentialArchetype *PAT,
-      continue;
+    for (auto req : reqSig->getRequirements()) {
+      RequirementSource InnerSource(Kind, Source.getLoc());
+      addRequirement(req, InnerSource, T, Visited);
```
Ahhh, much nicer!

DougGregor commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -1013,52 +1076,66 @@ bool ArchetypeBuilder::addConformanceRequirement(PotentialArchetype *PAT,
+        auto AssocPA = T->getNestedType(AssocType, *this);
+
+        if (AssocPA != T) {
+          if (addAbstractTypeParamRequirements(AssocType, AssocPA, Kind,
```
`addAbstractTypeParamRequirements` has a hack that could probably go away now:

  // If this is an associated type that already has an archetype assigned,
  // use that information.
  auto *genericEnv = decl->getDeclContext()
      ->getGenericEnvironmentOfContext();
  if (isa<AssociatedTypeDecl>(decl) && genericEnv != nullptr) {

because the requirement signature should already handle all of the cases where this would kick in.

DougGregor reviewed (approved):
Very cool! Aside from needing to bump the module format version number and a few minor things, this looks great.

huonw committed:
[AST] PotentialArchetype's ConcreteTypeSource is in the Representative.

The correct/canonical ConcreteType and ConcreteTypeSource are both
stored in the Representative, but previously only the getter for the
former was tranversing the chain.

slavapestov commented on lib/AST/ArchetypeBuilder.cpp:
```diff
@@ -499,6 +499,45 @@ auto ArchetypeBuilder::PotentialArchetype::getArchetypeAnchor()
 
+// Give a nested type the appropriately resolved concrete type, based off a
+// parent PA that has a concrete type.
+static void concretizeNestedTypeFromConcreteParent(
```
What is this for?

huonw committed:
[AST] Diagnose recursive requirements whenever they appear.

Previously, these were only diagnosed in some situations, but the
compiler is becoming more flexible about when things get fed into
different places, and so can crop up elsewhere.

huonw committed:
[Sema] Conformances only need to be verified as used once.

huonw commented:
Addressed the review, but I added two extra commits. You might have some thoughts on second last one in particular, it feels a bit ugly to me.

DougGregor commented:
I'm okay the last two commits; I do agree that the second-to-last commit for marking recursive archetypes is ugly, but it's diagnosing something that will cease to be an error in the near-ish future.

huonw commented:
Indeed.

@swift-ci Please test and merge

DougGregor commented:
w00t!
